addon-chrome/ChromePlacesProvider.js

```diff
@@ -0,0 +1,283 @@
+	}
+
+	static getHighlightsImg(sites) {
+		const imageWidth = 450;
+		const imageHeight = 278;
+
+		const highlightImgPromise = new Promise((resolve, reject) => {
+			const imgPromises = [];
+			for (let i = 0; i < sites.length; i++) {
+				let site = sites[i];
+				if (site.images) {
+					imgPromises.push(site);
+					continue;
+				}
+				const imgPromise = fetch(site.url)
+					.then((r) => r.text())
+					.catch((ex) => imgPromises.push(site)) // can't preview sites like localhost
+					.then((r) => {
+						const pageMetadata = getMetadata(new DOMParser().parseFromString(r, "text/html"));
+
+						const imageUrl = pageMetadata.image_url;
+						const description = pageMetadata.description;
+						const images = [];
+
+						if (imageUrl) {
+							images.push({
+								url: imageUrl,
+								width: imageWidth,
+								height: imageHeight
+							});
+						}
+
+						Object.assign(site, {images, description});
+
+						return site;
+					});
+
+					imgPromises.push(imgPromise);
+			}
+
+			Promise.all(imgPromises).then((highlights) => {
+				highlights.forEach((highlight) => {
+					if (!highlight) return;
+					if (highlight.dateAdded) {
+						db.addToDb(BOOKMARK, highlight);
+					}
+					if (highlight.lastVisitTime) {
+						db.addToDb(HISTORY, highlight);
+					}
+				});
+				resolve(highlights);
+			});
+		});
+
+		return highlightImgPromise;
+	}
+
+	static _getBookmark() {
+		const bookmarkPromise = new Promise((resolve, reject) => {
+			chrome.bookmarks.getTree((trees) => {
+				const rows = this._processBookmarks(trees)
+					.filter((bookmark) => !!bookmark.url);
+				resolve(rows);
+			});
+		});
+
+		return bookmarkPromise;
+	}
+
+	static _getHistory(options) {
+		const defaultOption = {
+			text: "",
+			startTime: this._getLastRequestTime() || new Date().getTime() - (7 * 24 * 60 * 60 * 1000)
+		};
+		const searchOptions = options ? Object.assign(defaultOption, options) : defaultOption;
+		const startTime = searchOptions.startTime;
+		const endTime = searchOptions.endTime;
+		const historyPromise = new Promise((resolve, reject) => {
+			chrome.history.search(searchOptions, (histories) => {
+				if (startTime && endTime) {
+					// api uses start and end time as OR instead of AND
+					// so filter it for now or we could just use this ???
+					histories = histories.filter((result) => result.lastVisitTime > startTime && result.lastVisitTime < endTime);
+				}
+				const rows = histories.map((result) => this.transformHistory(result));
+				resolve(rows);
+			});
+		});
+
+		return historyPromise;
+	}
+
+	static _setLastRequestTime(time) {
+		window.localStorage.setItem("lastrequestdate", time);
+	}
+
+	static _getLastRequestTime() {
+		return parseInt(window.localStorage.getItem("lastrequestdate"), 10);
+	}
+
+	static _filterBlockedUrls(items, blockedUrls) {
+		return items.filter((item) =>
+			blockedUrls.map((blocked) => blocked.url).indexOf(item.url) === -1);
+	}
+
+	static _mergeLinks(r1, r2) {
```

Comment:
This is actually merging bookmark GUIDs from r2 (assuming r2 is a list of bookmarks?) into r1. Consider renaming this method and variable names to make this explicit. Also, method comments.